### DISCUSSION TITLE
Building libcurl on ARM64 windows

### DISCUSSION THREAD

**Discussion by Bab95** [curl/curl#10663] [Q&A]
How to build libcurl for ARM64 architecture. 
I tried to build it in an ARM64 based Azure VM with windows11 using same steps i did for x64/x86
in addition added step MACHINE=arm64 as mentioned in build steps.
But when i took dumpbin it seems to building for x86 (which i believe is default build configuration.

I want to understand is it possible to build for ARM64 windows platform using visual studio?
What extra steps may i require to do so? 
The goal is to build static curl library for ARM64 windows platform.

**Comment by MarcelRaad** [comment]
What build system are you using? I'm successfully doing that using CMake's Visual Studio generator.

**Reply by Bab95** [reply]
I have tried on ARM64 based machine but it fails everytime with the same error.
Looks like it is trying to build x86 even when MACHINE=arm64 is provided.
i used command >nmake /f Makefile.vc VC=17 mode=static DEBUG=no MACHINE=arm64
I am not sure if this was the right way to build it but in winbuild readme file MACHINE option was mentioned. I though it would be easy to build it as for x64 it worked fine.

![image](https://user-images.githubusercontent.com/22617997/222702421-99077ac7-106c-4156-9a8f-9d671b678428.png)



@MarcelRaad  can you share the steps you took to build it?
What all tools do i need to install?

**Reply by MarcelRaad** [reply]
From a Visual Studio command prompt in the repo root:
```
cmake -A ARM64 -B ..\curl-build
cmake --build ..\curl-build
```

**Reply by jay** [reply]
> I am not sure if this was the right way to build it but in winbuild readme file MACHINE option was mentioned. I though it would be easy to build it as for x64 it worked fine.

In addition to setting MACHINE you must use the proper build environment for the architecture you are targeting. That said I don't know why it says x86. It looks like you have the build environment set to cross compile for x64. You will need to open the prompt that sets the build environment for ARM64. Try searching ARM64 Tools Command Prompt.

**Reply by Bab95** [reply]
I have added the screenshot of the command line tool I used to trigger the build. But I tried the cmake method suggested above seems to be working. Although I haven't tested the binaries yet.

In cmake build I am facing one problem that it's always building in debug mode. I am bit clumsy with cmake files. Although I was able to create release one as well by replacing CMAKE_C_FLAGSS_DEBUG with release one. But still I would like to know the exact configuration to set the release build.


One more question I have here is that the method above is not adding include/curl folder. Is it the same expected behaviour?

**Reply by MarcelRaad** [reply]
> But still I would like to know the exact configuration to set the release build.

You can choose a configuration when building with:
```
cmake --build --config <Debug | Release | RelWithDebInfo | MinSizeRel>
```

**Reply by jay** [reply]
Try doing it like this
cd builddir
cmake ..\curldir -DCURL_USE_SCHANNEL=ON -DCMAKE_INSTALL_PREFIX:PATH=C:\foo\bar
cmake --build . --target INSTALL --config Release

**Comment by Bab95** [comment]
libcurl built using above method always give unsupported protocol error with curl error code **1**
ssl is supported as i checked it using following snippet.
```
    curl_version_info_data* vinfo = curl_version_info(CURLVERSION_NOW);
    if (vinfo->features && CURL_VERSION_SSL) {
        std::cout << "SSL supported!" << std::endl;
    }
```

**Reply by jay** [reply]
~~How is that related to this discussion?~~ If you are reporting a bug please open an [issue](https://github.com/curl/curl/issues/new?assignees=&labels=&template=bug_report.md&title=) and include a minimal self contained example that can be used to reproduce.
